app/core/src/main/resources/static/js/pages/edit-table-of-contents.js

```diff
@@ -613,6 +618,49 @@ document.addEventListener("DOMContentLoaded", function () {
     updateEmptyStateButton();
   }
 
+  // Undo button funcionality
+  const undoBtn = document.getElementById("undoBtn");
+
+  function saveBookmarkState() {
+    previousBookmarks = [...bookmarks];
```

Comment:
copies the array, not the nested bookmark objects right?
If you change a bookmark object after saving, undo won’t fully revert it ithink